### DISCUSSION TITLE
SK6812 with an ESP32: white channel calibration for HyperSPI and HyperSerialWLED?

### DISCUSSION THREAD

**Discussion by pimw1** [awawa-dev/HyperHDR#246] [Q&A] [answered]
Hi,

I see that the white channel (SK6812 Neutral White) calibration is offered for HyperSerialESP32. Is it somehow possible to utilize the same color profiles for HyperSPI and HyperSerialWLED?

At this point, i would like to keep all my options open (especially since i am using the awesome Quinled ESP32 (<https://quinled.info/quinled-esp32/>) with ethernet board attached (<https://quinled.info/quinled-esp32-ethernet/>). A very nice setup, but also a setup for which i should take extra notion of which pins (not) to use.

Regards,

Pim

**Comment by awawa-dev** [comment]
Hi
There is a version of HyperSPI for neutral (and cold) white which is using the same algorithm as HyperSerialESP32, HyperSerialWLED is using custom WLED algorithms. Personally I wasn't satisfied with WLED color calibration because "neutral" white (I wonder who chose that name for the market?) is too yellow and WLED doesn't improve the situation much.  
  
So you can have HyperSerialWLED on one side and HyperSerialEsp8266/HyperSerialEsp32/HyperSpi (they share same two color profiles: one for cold and one for neutral white) on the other side. Just choose what suits you best, it is an individual matter.  
  
You can also use cold version firmware for "neutral" SK6812 and see the effect, because cold firmware treats each color channel equally for the white channel calculation (and that's the REAL neutral white).
Best regards
Awawa

**Comment by pimw1** [comment]
Thanks awawa-dev! I'll check it out this weekend and report back.

**Comment by pimw1** [comment]
Hi, a follow-up question: the ESP32 will be connected to a QuinLED-Dig-Quad, which is expecting data output to either GPIO16 / GPIO_3 (RX) / GPIO_1 (TX) / GPIO_4 .
![image](https://user-images.githubusercontent.com/74881023/158861475-5fdcdef0-9ac2-4d54-b4af-fdf71767071a.png)

HyperSPI is pushing data out towards GPIO 2. Is it possible to reconfiuger this to GPIO16 for example?

**Comment by pimw1** [comment]
I see that the pin number is a setting. However, i am an absolute noob when it comes to compiling: i've never compiled anything in my life. Could you help out a little bit?
![image](https://user-images.githubusercontent.com/74881023/158867961-9006dd6e-8209-42a0-827b-61a8a7cb062a.png)

**Comment by awawa-dev** [comment]
First prepare your Arduino IDE: https://randomnerdtutorials.com/installing-the-esp32-board-in-arduino-ide-windows-instructions/
Then install Makuna Neopixel Bus library:
![obraz](https://user-images.githubusercontent.com/69086569/158869462-f0cc0a48-8325-4cb4-a72a-03813deb477a.png)
You have to install drivers for your ESP32 and it should be detected as COMx serial port which you set in Arduino IDE.
Above steps have nothing to do with compiling so you should succeed.
Then open ino project and change that DATA_PIN. If you have connected and detected ESP32 in your system, Arduino IDE should offer you to compile and upload the firmware directly to ESP32.

**Reply by pimw1** [reply]
Thank you, i think i'm almost there. I followed all steps. I'm stuck at  the step "Then open ino project and change that DATA_PIN.".

I cannot find the DATA_PIN: the ino file looks completely different in Arduino IDE as compared to in github

![image](https://user-images.githubusercontent.com/74881023/158873980-6aebbcae-a267-443b-9c53-38d69c4c94c9.png)

**Comment by awawa-dev** [comment]
That's not ino project. Download ZIP'ed project from github and extract that file.

**Comment by pimw1** [comment]
Aah, check, got it. So this should do the trick (for gpio16 for SK6812 NW)?

![image](https://user-images.githubusercontent.com/74881023/158875155-2dcdd5ff-2c76-413c-9f62-e31f0ac6eecf.png)

**Reply by awawa-dev** [reply]
Yes, but that's HyperSerialESP32 not HyperSPI you mentioned earlier if that counts.

**Comment by pimw1** [comment]
Yeah, i first wanna try with HyperSerialESP32, i think that will be my best option with the current setup. Thanks for your help. I'll continue this weekend and update. The whole project is getting pretty awesome: 144led/meter ambilight in a custom made case with big power supply and big wires for rock stable voltage. The setup is currently with 144 led/meter ws2812b with Hyperion NG -> ESP32 WLED, but i am going to experiment with 144 led/meter SK6812NW with HyperHDR -> HyperSerialESP32. I'm curious how much the quality gain will be.

![image](https://user-images.githubusercontent.com/74881023/158876083-4038cf25-83d7-44f4-bc6c-37769e8a7fea.png)
![image](https://user-images.githubusercontent.com/74881023/158876145-d7ecc7e9-b962-4643-aeca-24ede41fc813.png)

**Comment by pimw1** [comment]
Hi, i'm unable to compile. It says: 
exit status 1
'NeoEsp32I2s1800KbpsMethod' was not declared in this scope

![image](https://user-images.githubusercontent.com/74881023/159137093-7e837b56-1649-4520-b754-620bae6a55ea.png)

Do you have any idea what could cause this?

**Reply by awawa-dev** [reply]
Yes. You haven't installed required library that I mentioned.

**Comment by pimw1** [comment]
I have installed those libraries. I think i know what is going on: i should first select my board. The error is gone after that.

This stuff is pretty hard to do when you have zero experience with compiling and arduinos in general.

![image](https://user-images.githubusercontent.com/74881023/159137276-b7f178eb-8eb1-4cae-a9de-49cd90969bdb.png)

**Reply by awawa-dev** [reply]
Yes. That library won't initialize when incorrect board is selected also (so the first point from the previous todo list). It's hardware depend.

**Reply by pimw1** [reply]
So confusing without understanding the basics.... I'll struggle on for a bit and report back.

**Reply by awawa-dev** [reply]
You've solved it in only 7 minutes :) The rest is a piece of cake.

**Comment by pimw1** [comment]
Alright, flashing is done and i've created a very simple test setup (just one ledstrip). It seems to be working: HyperHDR is nicely pushing everything to the ESP32 using the ADA protocol at 2000000 baudrate. I'll now connect everything. So far so good.

**Comment by pimw1** [comment]
HyperSerialESP32 is not working for me: my ESP32 is most of the times not booting when connected to usb (most likely a design issue of this specific board).

Therefore, i'm switching to HyperSPI. Crimped all dupont cables, hardware-wise looking good. The only thing left, is to change the output pin from GPIO2 to GPIO16. How could i change this in the firmware of HyperSPI?

**Comment by pimw1** [comment]
Would this work?
![image](https://user-images.githubusercontent.com/74881023/159168477-6cfc0f21-0a8b-4d64-bcdb-97fe752e9d40.png)

**Comment by pimw1** [comment]
Sorry, error, i mean like this: would that work?

![image](https://user-images.githubusercontent.com/74881023/159168776-d4486a3c-5ec7-42e3-b266-944fd2cfb573.png)

**Reply by awawa-dev** [reply]
Probably not. Judging from that piece of code (not knowing the rest) that's not the HyperSPI for ESP32 but some kind of hybrid. Also that booting issue indicates hardware/flash/connection problem with the board and I would stop here with further experiments.

**Reply by pimw1** [reply]
I don't wanna stop know after putting all that effort in haha. I made a silly mistake: i've opened ino from the wrong folder (ESP8266). 
I've found the proper ino now, and it contains the data_pin config. I'm at home with covid, that does not help with the sharpness of my brain.

Regarding a flash problem: i pretty much ruled that out.by providing myself some proof: i've flashed HyperSerialWLED and it works fine and communicating via network wled protocol. But when i reboot with the usb cable inserted, the wled-interface is not reachable anymore. I see online that more people have the same issue with usb: it's a known design flaw in some of the ESP32 wemos boards. So all in all, i don't think this is due to a flashing error tbh.

**Reply by awawa-dev** [reply]
I also had similar problem with WLED as many other people and that's proves nothing other that WLED currently is a bloatware  ;)

**Comment by pimw1** [comment]
Next error: do you have any thoughts what could be wrong?

![image](https://user-images.githubusercontent.com/74881023/159170142-1f65dee8-7ea1-44df-89db-d3f8942e0ec8.png)

**Reply by awawa-dev** [reply]
https://github.com/hideakitai/ESP32DMASPI/issues/13

**Comment by pimw1** [comment]
Thanks, one bug down, one to go:

which values should i use here?
mosi_io_num = 23
miso_io_num = 19
sclk_io_num = 18
quadwp_io_num = ?
quadhd_io_num = ?
max_transfer_sz = ?

spi_bus_config_t bus_cfg {
.mosi_io_num = 13, // HSPI
.miso_io_num = 12, // HSPI
.sclk_io_num = 14, // HSPI
.quadwp_io_num = -1,
.quadhd_io_num = -1,
.max_transfer_sz = 4092, // default: 4092 if DMA enabled, SOC_SPI_MAXIMUM_BUFFER_SIZE if DMA disabled
.flags = SPICOMMON_BUSFLAG_SLAVE,
.intr_flags = ESP_INTR_FLAG_IRAM
};

**Comment by pimw1** [comment, marked as answer]
It's still throwing an error, i'm clearly doing something wrong here. Could you give me some pointers?

![image](https://user-images.githubusercontent.com/74881023/159171599-b6bb70dc-2bc6-4087-9de5-0720bed30404.png)

**Reply by awawa-dev** [reply]
You are doing it wrong, not as the developer of that library requires it:
![obraz](https://user-images.githubusercontent.com/69086569/159171740-8684eb78-730d-4d3c-a5c0-04a3e1291423.png)
Or downgrade that library from 0.2.0 ...

**Reply by pimw1** [reply]
Cool, it's compiling and succesfully flashing now. I'll connect the stuff now and report back (fingers crossed ;))

![image](https://user-images.githubusercontent.com/74881023/159172134-00737b18-6f20-4458-b8dc-dcf53374a7a8.png)

**Comment by pimw1** [comment]
Alright, it works perfectly. Thank you for your assistance. Without it, i would be stuck.
![image](https://user-images.githubusercontent.com/74881023/159178927-09adc514-a9fb-4545-9772-0a4f12135665.png)

**Comment by pimw1** [comment]
One question though: i'd like to reattach my ethernet hat, but it is using the same GPIO's as are used for the input for HyperSPI.
Is it possible to use GPIO14+34+33+35 for SPI select, SCK, MISO and MOSI?

**Reply by awawa-dev** [reply]
Using ESP32DMASPI there are two options:
HSPI = CS: 15, CLK: 14, MOSI: 13, MISO: 12 (didn't tested)
VSPI  = CS:  5, CLK: 18, MOSI: 23, MISO: 19 (HyperSPI default SPI pinout)

**Reply by pimw1** [reply]
Ok thanks. That rules out the ethernet board than. Which is ok; it's not like there is a direct use-case for me in the current setup.
Thanks again for your help. Should i close this issue?

**Reply by awawa-dev** [reply]
If you think you've received all the answers then you can mark it as answered.

**Reply by pimw1** [reply]
Lol, i don't see the "Mark as answer" button below your replies, so i just pick one my own comments as "Mark as answer" :)
